### PR TITLE
Adding overly permissive CORS policy detector

### PR DESCRIPTION
**Pull request by plr0man**
Detector reports overly permissive CORS policy that sets asterisk as the value of the Access-Control-Allow-Origin header, like in the following examples:
response.addHeader("Access-Control-Allow-Origin", "\*");
response.setHeader("Access-Control-Allow-Origin", "\*");

### CODE REVIEW

**Comment by h3xstream** [comment]
Good one !

File: plugin/src/main/java/com/h3xstream/findsecbugs/PermissiveCORSDetector.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+                INVOKEINTERFACE invoke = (INVOKEINTERFACE) inst;
+                String methodName = invoke.getMethodName(cpg);
+                String className = invoke.getClassName(cpg);
+
+                if (className.equals("javax.servlet.http.HttpServletResponse") &&
+                   (methodName.equals("addHeader") || methodName.equals("setHeader"))) {
+
+                    LDC ldc = ByteCode.getPrevInstruction(location.getHandle().getPrev(), LDC.class);
+                    if (ldc != null) {
+                        if ("Access-Control-Allow-Origin".equals(ldc.getValue(cpg)) &&
```
Comment (h3xstream) [inline plugin/src/main/java/com/h3xstream/findsecbugs/PermissiveCORSDetector.java]: I would also test for the same header with different case (`access-control-allow-origin`). 
http://stackoverflow.com/a/5259004/89769

File: plugin/src/main/java/com/h3xstream/findsecbugs/PermissiveCORSDetector.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+                            "*".equals(ByteCode.getConstantLDC(location.getHandle().getPrev(), cpg, String.class))) {
```
Comment (h3xstream) [inline plugin/src/main/java/com/h3xstream/findsecbugs/PermissiveCORSDetector.java]: I would also test for wildcard use as sub-domain. (replace `equals` by `contains`)
http://blog.portswigger.net/2016/10/exploiting-cors-misconfigurations-for.html

Also the exact value "null" could be detected

File: plugin/src/main/java/com/h3xstream/findsecbugs/PermissiveCORSDetector.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+            Instruction inst = location.getHandle().getInstruction();
+
+            if (inst instanceof INVOKEINTERFACE) {
+                INVOKEINTERFACE invoke = (INVOKEINTERFACE) inst;
+                String methodName = invoke.getMethodName(cpg);
+                String className = invoke.getClassName(cpg);
+
+                if (className.equals("javax.servlet.http.HttpServletResponse") &&
+                   (methodName.equals("addHeader") || methodName.equals("setHeader"))) {
```
Comment (h3xstream) [inline plugin/src/main/java/com/h3xstream/findsecbugs/PermissiveCORSDetector.java]: Bonus : Typically website that need to use `Access-Control-Allow-Origin` will analyze the `Origin` header.
http://blog.portswigger.net/2016/10/exploiting-cors-misconfigurations-for.html

**This is not mandatory for this PR.** The description will have to be guideline of how to analyse properly Origin header. I would be interested to write this one.

**Comment by ptamarit** [comment]
Maybe it would make sense to move this detector in a `cors` sub-package.

Not sure I'd have time to do it, but similar checks specific to Spring Framework could be added (see https://docs.spring.io/spring-security/site/docs/current/reference/html/cors.html and http://docs.spring.io/spring/docs/current/spring-framework-reference/html/cors.html).

File: plugin/src/main/java/com/h3xstream/findsecbugs/PermissiveCORSDetector.java (hunk 3 above)
Comment (plr0man) [inline plugin/src/main/java/com/h3xstream/findsecbugs/PermissiveCORSDetector.java]: Nice! Go ahead, I'm looking forward to seeing how you approach it.

File: plugin/src/main/java/com/h3xstream/findsecbugs/PermissiveCORSDetector.java [hunk 4]
```diff
@@ -81,8 +81,9 @@ private void analyzeMethod(Method m, ClassContext classContext) throws CFGBuilde
 
                     LDC ldc = ByteCode.getPrevInstruction(location.getHandle().getPrev(), LDC.class);
                     if (ldc != null) {
-                        if ("Access-Control-Allow-Origin".equals(ldc.getValue(cpg)) &&
-                            "*".equals(ByteCode.getConstantLDC(location.getHandle().getPrev(), cpg, String.class))) {
+                        String headerValue = ByteCode.getConstantLDC(location.getHandle().getPrev(), cpg, String.class);
+                        if ("Access-Control-Allow-Origin".equalsIgnoreCase((String)ldc.getValue(cpg)) &&
```
Comment (plr0man) [inline plugin/src/main/java/com/h3xstream/findsecbugs/PermissiveCORSDetector.java]: Fixed case sensitivity, header value wildcards and "null" value.

File: plugin/src/main/java/com/h3xstream/findsecbugs/PermissiveCORSDetector.java [hunk 5]
```diff
@@ -81,8 +81,9 @@ private void analyzeMethod(Method m, ClassContext classContext) throws CFGBuilde
...
+                            (headerValue.contains("*") || "null".equalsIgnoreCase(headerValue))) {
```
Comment (plr0man) [inline plugin/src/main/java/com/h3xstream/findsecbugs/PermissiveCORSDetector.java]: Fixed case sensitivity, header value wildcards and "null" value.

**Pull request merged**